libmamba/src/core/download_impl.hpp

```diff
@@ -47,43 +49,60 @@ namespace mamba
-        static size_t curl_write_callback(char* buffer, size_t size, size_t nbitems, void* self);
-        static int curl_progress_callback(
-            void* f,
-            curl_off_t total_to_download,
-            curl_off_t now_downloaded,
-            curl_off_t,
-            curl_off_t
-        );
-
-        bool can_retry(CURLcode code) const;
-        bool can_retry(const TransferData& data) const;
-
-        TransferData get_transfer_data() const;
-        DownloadError build_download_error(CURLcode code) const;
-        DownloadError build_download_error(TransferData data) const;
-        DownloadSuccess build_download_success(TransferData data) const;
-
-        const DownloadRequest* p_request;
-        CURLHandle m_handle;
-        on_success_callback m_success_callback;
-        on_failure_callback m_failure_callback;
-        std::size_t m_retry_wait_seconds;
-        std::unique_ptr<CompressionStream> p_stream;
-        std::ofstream m_file;
-        std::string m_cache_control;
-        std::string m_etag;
-        std::string m_last_modified;
+        // This internal structure stored in an std::unique_ptr is required to guarantee
+        // move semantics: some of these functions return lambda capturing the current
+        // instance, therefore this latter must be stable in memory.
+        struct Impl
+        {
+            Impl(
+                CURLHandle& handle,
+                const DownloadRequest& request,
+                CURLMultiHandle& downloader,
+                const Context& context,
+                on_success_callback success,
+                on_failure_callback error
+            );
+
+            bool finish_download(CURLMultiHandle& downloader, CURLcode code);
+            void clean_attempt(CURLMultiHandle& downloader, bool erase_downloaded);
+            void invoke_progress_callback(const DownloadEvent&) const;
+
+            void configure_handle(const Context& context);
+            void configure_handle_headers(const Context& context);
+
+            size_t write_data(char* buffer, size_t data);
+
+            static size_t curl_header_callback(char* buffer, size_t size, size_t nbitems, void* self);
+            static size_t curl_write_callback(char* buffer, size_t size, size_t nbitems, void* self);
+            static int curl_progress_callback(
+                void* f,
+                curl_off_t total_to_download,
+                curl_off_t now_downloaded,
+                curl_off_t,
+                curl_off_t
+            );
+
+            bool can_retry(CURLcode code) const;
+            bool can_retry(const TransferData& data) const;
+
+            TransferData get_transfer_data() const;
+            DownloadError build_download_error(CURLcode code) const;
+            DownloadError build_download_error(TransferData data) const;
+            DownloadSuccess build_download_success(TransferData data) const;
+
+            CURLHandle* p_handle = nullptr;
```

Comment:
Isnt handle already a value/pointer? Do you need to point to it?

Reply:
With this change, the handle is now a member (value) of DownloadTracker, to avoid recreating it each time we build an attempt. BUt the attempt still needs it, thus this pointer.